### PR TITLE
Fixed #26521 - CreateModel allows duplicate field names

### PR DESCRIPTION
**Pull request by jiaaro**
I did this as part of the sprint tonight at Django NYC cc/ @charettes 

https://code.djangoproject.com/ticket/26521

### CODE REVIEW

**Comment by w0rp** [comment]
Looks pretty good to me, so long as it passes all of the required conventions and such.

**Comment by charettes** [comment]
Buildbot, add to whitelist.

**Comment by MarkusH** [comment]
If we're already doing this, can we check for manager names and uniqueness of bases, too. please.

**Comment by jiaaro** [comment]
@MarkusH no problem :)

While adding checks for uniqueness of bases I discovered that some of the tests optimizer tests were not passing in iterables. Since there are numerous places in the `CreateModel` code that iterate over `self.bases` I assumed this was an oversight and added the missing commas.

Still, I wanted to mention it - I could be wrong!

File: django/db/migrations/operations/models.py [hunk 1]
```diff
@@ -42,8 +42,38 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
...
         super(CreateModel, self).__init__(name)
 
+        # Sanity-check that there are no duplicated field names, bases, or
+        # manager names
+        used_field_names = set()
+        for name, _ in fields:
+            if name in used_field_names:
+                raise ValueError(
+                    'You cannot have more than one field with the same name. '
+                    '"%s" was used more than once' % name
```
Comment (charettes) [inline django/db/migrations/operations/models.py]: Add a trailing period.

File: django/db/migrations/operations/models.py [hunk 2]
```diff
@@ -42,8 +42,38 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
...
+                    '"%s" was used more than once.' % base
+                )
+            used_bases.add(base)
+
+        used_manager_names = set()
+        for name, _ in self.managers:
+            if name in used_manager_names:
+                raise ValueError(
+                    'You cannot have more than one field with the same name. '
+                    '"%s" was used more than once' % name
```
Comment (charettes) [inline django/db/migrations/operations/models.py]: Ditto.

File: django/db/migrations/operations/models.py [hunk 3]
```diff
@@ -42,8 +42,38 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
...
+        for name, _ in fields:
+            if name in used_field_names:
+                raise ValueError(
+                    'You cannot have more than one field with the same name. '
+                    '"%s" was used more than once' % name
+                )
+            used_field_names.add(name)
+
+        used_bases = set()
+        for base in tuple(self.bases):
```
Comment (charettes) [inline django/db/migrations/operations/models.py]: I think we want to crash here if `bases` is not iterable.

File: django/db/migrations/operations/models.py [hunk 4]
```diff
@@ -42,8 +42,38 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
...
+
         super(CreateModel, self).__init__(name)
 
+        # Sanity-check that there are no duplicated field names, bases, or
+        # manager names
+        used_field_names = set()
+        for name, _ in fields:
+            if name in used_field_names:
+                raise ValueError(
+                    'You cannot have more than one field with the same name. '
```
Comment (timgraham) [inline django/db/migrations/operations/models.py]: How about more simply: "Found duplicate <field/base/manager> '%s' in CreateModel operation."
Maybe you could create a helper function so we don't have to repeat a similar loop 3 times.

File: tests/migrations/test_operations.py [hunk 5]
```diff
@@ -199,6 +199,42 @@ def test_create_model(self):
         definition = operation.deconstruct()
         self.assertNotIn('managers', definition[2])
 
+    def test_create_model_with_duplicate_field_name(self):
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: Please use `self.assertRaisesMessage` to verify this is the expected `ValueError` .

File: django/db/migrations/operations/models.py (hunk 3 above)
Comment (jiaaro) [inline django/db/migrations/operations/models.py]: we still do - tuple requires an iterable, though you're right, there's no reason to use `tuple()` anymore

Comment (charettes) [inline django/db/migrations/operations/models.py]: right, I missed that :)

File: django/db/migrations/operations/models.py [hunk 6]
```diff
@@ -42,8 +42,38 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
         self.options = options or {}
         self.bases = bases or (models.Model,)
         self.managers = managers or []
+
```
Comment (charettes) [inline django/db/migrations/operations/models.py]: Revert this new line addition as it makes blaming harder.

File: django/db/migrations/operations/models.py (hunk 4 above)
Comment (jiaaro) [inline django/db/migrations/operations/models.py]: @timgraham I added the helper function to the module - is there a better place for it?

File: tests/migrations/test_optimizer.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def test_create_rename_model(self):
                     name="Foo",
                     fields=[("name", models.CharField(max_length=255))],
                     options={'verbose_name': 'Foo'},
-                    bases=(CustomModelBase),
+                    bases=(CustomModelBase,),
```
Comment (MarkusH) [inline tests/migrations/test_optimizer.py]: Nicely spotted!

File: tests/migrations/test_operations.py [hunk 8]
```diff
@@ -199,6 +199,42 @@ def test_create_model(self):
...
+            migrations.CreateModel(
+                "Pony",
+                fields=[],
+                bases=("test_crmo.Pony", "test_crmo.Pony",),
+            )
+        with self.assertRaisesMessage(ValueError, "Found duplicate base in CreateModel operation."):
+            migrations.CreateModel(
+                "Pony",
+                fields=[],
+                bases=(CustomModelBase, CustomModelBase,),
```
Comment (charettes) [inline tests/migrations/test_operations.py]: The `CustomModelBase` class is not an actual `models.Model` subclass but a `models.base.ModelBase` subclass which is `type(models.Model)` (`models.Model`'s metaclass). We should use an actual `models.Model` subclass instead.

File: django/db/migrations/operations/models.py [hunk 9]
```diff
@@ -43,6 +53,11 @@ def __init__(self, name, fields, options=None, bases=None, managers=None):
         self.bases = bases or (models.Model,)
         self.managers = managers or []
         super(CreateModel, self).__init__(name)
+        # Sanity-check that there are no duplicated field names, bases, or
+        # manager names
+        check_for_duplicates((name for name, _ in self.fields), "field")
+        check_for_duplicates(self.bases, "base")
```
Comment (charettes) [inline django/db/migrations/operations/models.py]: Check against `(base._meta.label_lower if isinstance(base, models.base.ModelBase) else base.lower() for base in self.bases)` to catch mixed types bases duplicates (e.g. `bases=(MyModel, MyModel._meta.label_lower)`)

File: tests/migrations/test_operations.py [hunk 10]
```diff
@@ -199,6 +199,42 @@ def test_create_model(self):
...
+            )
```
Comment (charettes) [inline tests/migrations/test_operations.py]: Add tests for `bases=('test_crmo.Pony', 'test_crmo.pony')` duplicates (case mismatch) and the mixed type case described above.

File: django/db/migrations/operations/models.py (hunk 9 above)
Comment (jiaaro) [inline django/db/migrations/operations/models.py]: @charettes in many cases `models.Model` is used as the base (it's the default) and that does not have a `_meta` attribute - is this something that can safely be special-cased? or could it be referenced by a dotted-path string (e.g., 'django.db.models.Model')?

Comment (charettes) [inline django/db/migrations/operations/models.py]: I see. Adding an `if base is not models.Model` to the generator should do.

**Comment by jiaaro** [comment]
@charettes I think/hope I've addressed all your comments

**Comment by charettes** [comment]
Thanks for everything @jiaaro.

Do you think you could squash all your commits into a single one and give it a message reflecting all the changes it represents?

**Comment by jiaaro** [comment]
@charettes good?

**Comment by charettes** [comment]
Just merged it, thanks!

File: django/db/migrations/operations/models.py [hunk 11]
```diff
@@ -12,6 +12,16 @@
 )
 
 
+def _check_for_duplicates(arg_name, objs):
+    used_vals = set()
+    for val in objs:
+        if val in used_vals:
+            raise ValueError(
+                'Found duplicate %s in CreateModel operation.' % arg_name
```
Comment (timgraham) [inline django/db/migrations/operations/models.py]: I didn't follow the discussion closely but was there a reason for not including the duplicate value in the message? I thought that was in the first version of the message.

Comment (charettes) [inline django/db/migrations/operations/models.py]: It was in the initial version of patch when bases and managers were not also checked. If you think this must be present I can submit a PR with the previous changes.

Comment (timgraham) [inline django/db/migrations/operations/models.py]: Is there any more complication than adding `val` in the string interpolation: `duplicate %s (%s)`? Just seems like it's a trivial addition without any downside.

Comment (charettes) [inline django/db/migrations/operations/models.py]: Right, I'll sumit a PR shortly with the required changes.

**Pull request merged**